src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,142 @@
+      Function(name, args, results, pres, posts, body map (computeNewBody(args, _)))(f.info)
+
+    // same as functions
+    case m@Method(receiver, name, args, results, pres, posts, body) =>
+      Method(receiver, name, args, results, pres, posts, body map (computeNewBody(args, _)))(m.info)
+
+    // Adds pre-conditions stating the bounds of each argument and a post-condition to check if the body expression
+    // overflows
+    case f@PureFunction(name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPre(args)
+        val newPost = posts ++ Vector(getPureBlockPost(expr, results))
+        PureFunction(name, args, results, newPre, newPost, body)(f.info)
+      case None => f
+    }
+
+    // Same as pure functions
+    case m@PureMethod(receiver, name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPre(args)
+        val newPost = posts ++ Vector(getPureBlockPost(expr, results))
+        PureMethod(receiver, name, args, results, newPre, newPost, body)(m.info)
+      case None => m
+    }
+
+    // TODO: should we add overflow checks to predicates?
+
+    case x => x
+  }
+
+  /**
+    * Adds overflow checks to the bodies of a block. If block is parametrized,
+    * `computeNewBody` adds assumptions about the parameters' bounds
+    */
+  private def computeNewBody(args: Vector[Parameter.In], body: Block): Block = {
+    val assumeArgBounds = args map { case p@Parameter.In(_, typ) => Assume(assertionExprInBounds(p, typ)(p.info))(p.info) }
+    val blockStmts = body.stmts map stmtTransform
+    Block(body.decls, assumeArgBounds ++ blockStmts)(body.info)
+  }
+
+  /**
+    * Computes the pre-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that each argument is within the bounds for its type
+    */
+  private def getPureBlockPre(args: Vector[Parameter.In]): Vector[Assertion] = {
+    args.map { p: Parameter.In => assertionExprInBounds(p, p.typ)(p.info) }
+  }
+
+  /**
+    * Computes the post-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that the expression result is within the bounds of its type
+    */
+  private def getPureBlockPost(body: Expr, results: Vector[Parameter.Out]): Assertion = {
+    // relies on the current assumption that pure functions and methods must have exactly one result argument
+    if (results.length != 1) violation("Pure functions and methods must have exactly one result argument")
+    assertionExprInBounds(body, results(0).typ)(addAnnotation(body.info))
+  }
+
+  private def stmtTransform(stmt: Stmt): Stmt = stmt match {
+    case b@Block(decls, stmts) => Block(decls, stmts map stmtTransform)(b.info)
+
+    case s@Seqn(stmts) => Seqn(stmts map stmtTransform)(s.info)
+
+    case i@If(cond, thn, els) =>
+      val condInfo = addAnnotation(cond.info)
```

Comment:
I don't understand what `addAnnotation` does / how it works. I would have assumed that you add some back-tracking information for the newly generated assertions but why is it called on the original condition?

Reply:
I indeed add an `OverflowCheckAnnotation` label to a node's `info` field. When an error is reported, I check whether the `info` field is marked with such an annotation and in such a case, I transform its error message from an assertion error to an overflow error. In this particular case, I'm adding an annotation to the branch condition because I am generating an overflow check for that expression and I want to be able to point to this expression if an overflow is detected here.

Reply:
Ah so this works because the generated assertion stmt and the node that causes this generation share the same `info`? 
Another consequence is that you couldn't distinguish which of the following two assertion failed and you would both translated to an overflow error:
```
assert 4 + 5 < MAX_INT // generated overflow check
assert 4 + 5 == 9 // original assertion
```
But this case is not an issue as (as discussed yesterday) no overflow checks should be produced for expressions in assertions, right?

Reply:
It is able to distinguish between the two cases because, on the first assertion, the assert statement would be "marked" with an `OverflowCheckAnnotation` and the origin information of the expression node that caused the generation of the assertion whereas the second one would not. As an demo, the program
```go
// Should overflow when running in 32 bit mode, not in 64
requires i <= 2147483647
ensures res == i + 1
func overflowInt32(i int) (res int) {
	assert 5 + 1 == 5
	return i + 1
}
```
leads to the error
```
[info] 	<28:2> Assert might fail. 
[info] Assertion 5 + 1 == 5 might not hold
```
but if we comment the line ```assert 5 + 1 == 5``` then it produces the error
```
[info] 	<29:9> Expression may cause integer overflow. 
[info] Expression i + 1 may cause integer overflow
```

Reply:
Regarding the second observation about not producing overflow checks for assertions, this is achieved by mapping Assert, Assume, Inhale and Exhale statements to themselves in the program transformation that adds Overflow Checks

Reply:
> It is able to distinguish between the two cases because
Are you sure? I meant a program that only contains the assertion stmt `assert 4 + 5 == 9` which would then result in generating the other assert stmt (to check for a potential overflow). My understanding so far is that you call `addAnnotation` on the original assert stmt and that this works because both assert stmts share the same `info` object. If my thoughts are correct, then an 'assertion might not hold' error is transformed to an overflow error for both assert stmts

Reply:
The info field is not necessarily shared. When I call `addAnotation`, I create a new `Info` value that contains the previous one but is annotated with a label to identify it as an overflow check. I then use it as the info field only on expressions that check for overflows

Reply:
The info field is not necessarily shared. When I call `addAnotation`, I create a new Info value that contains the previous one but is annotated with a label to identify it as an overflow check. I then use it as the info field only on expressions that check for overflows

Reply:
Ah! I totally missed that `addAnnotation` returns a new info object. I thought you're attaching some information to the existing info object

Reply:
I will change the name `addAnnotation` to `createAnnotatedInfo` to make it clearer